### ISSUE TITLE
Fix getifaddrs syscall

### ISSUE THREAD

**Issue by qianxichen233** [Lind-Project/lind-wasm#20]
The getifaddrs syscall is not working in lind-wasm since the memory allocation to ifaddrs struct seems to be wrong. The getifaddrs syscall takes a double pointers (struct ifaddrs **ifap), and it is underlying syscall that is responsible for allocating the memory for the ifap struct. In rawposix, getifaddrs_syscall treats  the ifap as an one-dimensional pointer (*mut u8), and fill the pointer with the ifaddrs from libc. The correct steps should be something like:
```
*buf = malloc_wasm_memory(...)
fill(*buf, &ifap_from_libc)
... (malloc_wasm_memory for next ifap struct)
```
The main difficuly with this approach is that, it has to malloc the wasm memory inside rawposix, but the wasm memory is managed by glibc, so it would be kind tricky to let rawposix malloc the wasm memory.

Another approach would be using the ```getifaddrs_internal``` function provided by glibc. This function does not actually use getifaddrs syscall directly, but instead used some other network syscalls to simulate the getifaddrs. Currently this approach would fail since it used bind syscall on socket type that is not directly supported by rawposix (AF_NETLINK socket). However, this is an issue in rawposix (https://github.com/Lind-Project/RawPOSIX/issues/3) and will be likely fixed soon.

**Comment by qianxichen233** [comment]
Looks like `getifaddrs` is not a syscall but a glibc function. But we somehow has a `getifaddrs_syscall` in rawposix. This should be deleted I guess.

Turns out `getifaddrs` in glibc is internally using some socket types that we currently do not support, so I guess this issue is strongly associated with issue #42. Getting #42 fixed will probably also fix this issue

**Comment by robinyuan1002** [comment]
Yeah, seem like that getifaddrs if a function and it may call socket, ioctl syscall. Should I Delete this syscall first?

**Comment by rennergade** [comment]
Ah that makes sense. We handled getiffaddrs like this because we wanted to circumvent the weirder syscalls for popular paths.

Can we get an inventory of what syscalls getifaddrs uses and what kind of socket types? @robinyuan1002 

Also might make sense for you to work on #42 next so we can get to this.

**Comment by rennergade** [comment]
> Yeah, seem like that getifaddrs if a function and it may call socket, ioctl syscall. Should I Delete this syscall first?

If you want to put in a PR just removing getifaddrs as is from RawPOSIX for now that would be fine as well.

**Comment by robinyuan1002** [comment]
Ok, I will submit a PR

**Comment by Yaxuan-w** [comment]
I briefly looked the expansion of [getifaddrs](https://github.com/Lind-Project/lind-wasm/blob/main/src/glibc/sysdeps/unix/sysv/linux/ifaddrs.c) and it appears that glibc performs AF_NETLINK operations via the following sequence:

```
socket(AF_NETLINK, ...)
bind(sock, ...)
send(sock, ...)
recv(sock, ...)
```

Currently, if rawposix is not functioning as expected, it is most likely due to the fact that the returned socket domain should be `AF_NETLINK` (which is different from `AF_INET` (IPv4), `AF_INET6` (IPv6), or `AF_UNIX`). However, [in `bind_syscall`'s type conversion](https://github.com/Lind-Project/lind-wasm/blob/main/src/RawPOSIX/src/interface/types.rs#L422), we have not explicitly handled this case.

Previously, in the NaCl runtime environment, we explicitly handled different socket types independently. However, in lind-wasm, the implementation logic has changed, and rawposix now primarily forwards syscalls and ensures type alignment, rather than independently managing different socket types.

Resolving this issue should be possible by restructuring the data structures as proposed in #42.
